### ISSUE TITLE
[Spell] Death Knight - Death Rune Grace Period incorrect

### ISSUE THREAD

**Issue by festerhowl** [WoWManiaUK/Redemption#5337]
**What is rune grace period**: DK runes have a 10 sec cooldown when they are out of combat. When they go in combat, rune grace period activates, which means - once a used rune has replenished fully when you are in combat, it will continue to replenish that particular rune for an additional 2.5 seconds, such that the next time when you use that rune it will start counting down from 7.5 seconds instead of 10.

**What is Happening**: The rune grace period is not kicking in when you use death runes. It is working fine when you use the  other runes.

**how to get death runes**:   https://wotlk.evowow.com/?spell=50034  - Blood(**Death Rune Mastery**)
                                               https://wotlk.evowow.com/?spell=54637  -  Frost(**Blood of the North**)
                                               https://wotlk.evowow.com/?spell=56835  -  Unholy(**Reaping**)

**how to reproduce the issue**: while you are in combat, use any ability that costs a blood/frost/unholy rune and wait for that rune to replenish. 3 seconds after that particular rune has replenished, you will notice that when you use that rune again, it does not start counting down from 10 seconds but instead it has a lower cooldown. But when you try to use a death rune for the same thing, even after waiting 3 seconds, it starts counting down that rune from 10 instead of 7.5.

**Comment by Kilju** [comment]
I am no way an expert of DKs but this intrigued me so I investigated and tested out on my own a little on rune grace period. My proofable knowledge largely comes out of these links. Take this with a grain of salt as none of this is confirmed and is not 100% reliable data. Note that I generally dislike death knights quite a bit (Felt that it was fair to give warning that text might be biased)
My post is to supply information about rune grace period and its mechanics. 1 part is inconsistency with rune grace period in wow-mania, second part is possible explanation to behavior faced by festerhowl to help testers replicate it as I felt festerhowls explanation left room for more indepth explanation

https://www.mmo-champion.com/threads/713253-Question-about-rune-regen
https://web.archive.org/web/20120617161527/http://elitistjerks.com/f72/t94207-dk_mechanics_discussion/p2/
https://www.wowinterface.com/downloads/info14510-SnowfallRune.html#info

**1st part**
Rune grace period is designed to help death knights fit in runic power / non-rune abilities and not miss out damage in their rotation. I tested how rune grace period works on wow-mania. It seems like the grace period is 5 seconds here to get the benefit of -2.5seconds on a refreshed rune. When the intended mechanic is that you have to use that particular rune inside that 2,5 seconds from my understanding to gain benefit of 10s - x where x is the amount of time lost on the next refresh (up to 2,5seconds). Runes are supposed to be on 10 second cooldown, right now its possible to fit in 3 attacks or 4 in unholy presence (in perfect ms, 4,5seconds can be delayed safely) when the intended behaviour is to just be able to use 1 attack (1,5global + 200ms server ticks that blizzard servers had at that era (Reason why classic wow has "spell batching", to emulate the servers of the era) + player lag = 2,5 seconds to avoid weird server behavior / Players benefit)

Judging from snowfallrune addon the criterias for refreshed runes are supposed to be (Exact copypaste from the site):
In case some folks considering this addon don't know what the rune grace period is, let me give a brief explanation. For every death knight, a particular rune becomes available for use at the time when both of these conditions are satisfied:

    10 seconds after the rune last became available
    7.5 seconds after the rune was last used

The net result of this is that you'll be able to use each rune on average once every 10 seconds, and even if you don't use your runes as soon as they're available, your rune throughput isn't harmed in any way as long as you use your runes within 2.5 seconds after they're available. This 2.5 seconds is called the rune grace period.

**TL;DR version:** PVE implications: it is possible to have more forgiving rotation in PVE than intended because rune grace period is ~5 seconds, giving damage edge of 1-2 extra death coils if you have enough RP.
PVP Implications: It is possible to have your runes get refreshed in 7,5 seconds instead of 10 seconds when getting kited / unable to use runes due to range for example.  + the PVE implication.

**2nd part**
when I was scouring the internets and google for information on how (Death) runes are supposed to work, I came across inside the elitistjerks link explanation into how and what runes are used as death runes.

"From the testing I've done, death runes now longer work as the linked article specifies for a couple of situations. Namely, in the case of Death Rune Mastery, Death Strike, Icy Touch, and Plague Strike will preferentially use the death runes made from their original type. In other words, if you have BBDDDD available and use Death Strike, it will leave you with BB-D-D, rather than BBDD-- as it once did. Icy Touch will use death runes made from frost runes, and Plague Strike ones from unholy runes. I've yet to test it with Obliterate, Scourge Strike, or Howling Blast, but I'd assume it works similarly for all double-rune abilities. Blood rune abilities will still preferentially use both unholy-death runes before the frost-death runes, though. If you have BBDDDD, the order of use would be 125634 on blood rune abilities.

The odd behavior of which runes get converted to death runes from Death Rune Mastery remains, though. If you have both frost runes available as death runes, but both unholy runes are unavailable, using Death Strike will use both frost runes. The left frost rune will remain a death rune. If one of the unholy runes will refresh as a normal unholy rune, it will instead refresh as a death rune, preferring the left unholy rune if both will refresh as normal unholy runes. If both unholy runes will refresh as death runes, both frost runes will remain death runes."

Could it be possible that you missed out on the rune grace period due to it using a rune that was refreshed over 2,5 seconds ago? (Well if first part holds true, 5seconds I guess ?)

**Comment by festerhowl** [comment]
>I tested how rune grace period works on wow-mania. It seems like the grace period is 5 seconds here to get the benefit of -2.5seconds on a refreshed rune. When the intended mechanic is that you have to use that particular rune inside that 2,5 seconds from my understanding to gain benefit of 10s - x where x is the amount of time lost on the next refresh (up to 2,5seconds)
>Could it be possible that you missed out on the rune grace period due to it using a rune that was refreshed over 2,5 seconds ago? (Well if first part holds true, 5seconds I guess ?)

the grace period lasts indefinitely as long as you are in combat. If you had a limited time frame to use that rune, for instance 5 seconds for the 2.5 second benefit, the rune grace period's main purpose would diminish. The only way to reset the grace period is if you leave combat. I checked if the grace was indeed 5 seconds here but thankfully its not as that would've been another bug then :p


>PVE implications: it is possible to have more forgiving rotation in PVE than intended because rune grace period is ~5 seconds, giving damage edge of 1-2 extra death coils if you have enough RP

i am unsure about unholy but it definitely hurts frost dk in raids, because ideally you would want to use your blood death runes for an extra obliterate or a howling blast for AoE, but you end up using a pestilence and blood strike instead because the diseases run out before your blood runes can replenish. So right now the number of times you pestilence+blood strike your target divided by 2 is the number of obli/hb you are currently losing in an entire fight.


>PVP Implications: It is possible to have your runes get refreshed in 7,5 seconds instead of 10 seconds when getting kited / unable to use runes due to range for example. + the PVE implication.

i personally don't think it affects pvp all that much, because you go in and out of combat all the time so the grace period resets. It only matters when you are in a long duel and your opponent is facerolling you.

>Blood rune abilities will still preferentially use both unholy-death runes before the frost-death runes, though. If you have BBDDDD, the order of use would be 125634 on blood rune abilities.

this part really confused me because ideally the rune usage should be 123546, but for some reason blood rune abilities prefer unholy death runes to frost death runes.  At least it's a good thing blood is only good enough for tanking.

**Comment by Keader** [comment]
i realy need a tester to confirm it and get a good reproduce way.
Testing in debug here, everything looks fine o.o

**Comment by festerhowl** [comment]
basically the gist of the matter is that when you are in combat and try to convert death runes to blood/unholy/frost runes, the RGP does not affect it, so the converted runes always start their countdown from 10 seconds  even if you wait for a few seconds for the RGP to kick in. The RGP is working fine as long as you do not involve any death runes.

**Comment by svijetlo** [comment]
if you use Death rune to cast smt deth rune should be back after 7.5sec to be ready to use not 10 sec

**Comment by Keader** [comment]
can you give me a reproduce steps @svijetlo ?

**Comment by festerhowl** [comment]
- Get **Blood of the North 3/3** talent from frost tree OR **Reaping 3/3** from the unholy tree
- Get in combat
- Convert your blood runes to death runes via **pestilence/blood strike**
- wait for the death runes to finish their cooldown+ 2.5 seconds for maximum rune grace period(so 12.5 seconds in total)
- use your death runes.
_The blood runes generated from the usage of  death runes have a 10 second cooldown, when they should have less than that(down to 7.5 seconds depending upon how long you waited for the rune grace period). For example- if you waited 11 seconds instead of 12.5 seconds for the death rune to finish its cooldown, when you use the death rune, the blood runes generated in the process will have 9 second cooldown and so on_

**Comment by Keader** [comment]
test now pls

**Comment by hawker357** [comment]
Just checked this, as I understand it, whenever a Death Rune is used, if you wait more than 12.5 seconds before you use one, the CD for that specific rune should be 7.5 seconds. Regardless of how that Death Rune is created.

Currently on PTR the Death Rune CD is still behaving differently than the rest of the runes. Even after waiting ~ 15 seconds it still has the full 10 seconds CD.

**Comment by hawker357** [comment]
I'm still getting a full 10 sec CD on death runes on PTR

**Comment by hawker357** [comment]
Still needs some work on it, it appears that the cooldown for when a death rune is used is unchanged. The cooldown for when a death rune is created seems fine.

**Comment by Keader** [comment]
gime a reproduce steps, without it i cant do a shit.

**Comment by hawker357** [comment]
I used Death Strike with Improved Rune Mastery 3/3, which created 2 Death Runes, that specific cooldown was fine.
Then I spammed Icy Touch to use up the Death Runes, that specific cooldown still had the 10 seconds cooldown, even after the grace period.

**Comment by Keader** [comment]
test now pls

**Comment by hawker357** [comment]
Tested again using the same steps, Death Strike will not create Death Runes with Death Rune Mastery 3/3, also the CD on the Frost/Unholy runes used for the Death Strike is now ~5 seconds (too short).

**Comment by Keader** [comment]
I put back the version tested by @svijetlo
i cant change nothing more until i understand what is the problem now, because i cant reproduce.

**Comment by hawker357** [comment]
Okay I'll try to break down my understanding

Expected:
Pull a mob
Use Death Strike with 3/3 Death Rune Mastery
The runes used for Death Strike (1 Frost and 1 Unholy) will have a 7.5 second cooldown.
Death Runes are created in the place of the runes used for the Death Strike, with a 7.5 second cooldown.
Wait for Death Runes to charge + about 3 seconds.
Spam Icy Touch to use 1 Frost and 2 Death Runes, all 3 runes should have a 7.5 second cooldown.

Current:
Pull a mob
Use Death Strike with 3/3 Death Rune Mastery
The runes used for Death Strike (1 Frost and 1 Unholy) will have a 7.5 second cooldown.
Death Runes are created in the place of the runes used for the Death Strike, with a 7.5 second cooldown.
Wait for Death Runes to charge + about 3 seconds.
Spam Icy Touch to use 1 Frost and 2 Death Runes, the Frost rune has a 7.5 second cooldown, the 2 Death Runes have a 10 second cooldown.

I'm open for correction but this is my understanding

**Comment by festerhowl** [comment]
@Keader  There are **3 ways** to reproduce this issue. I am going to show all the ways i get my **Death Runes** along with the **talents** i used and the **rotation** and where the **Bug** is occuring

### Reproduce way 1(Threat of Thassarian):
_video link_- https://youtu.be/82_rv_MudxE

**Talents used:**
![Screenshot (182)](https://user-images.githubusercontent.com/43633844/111459946-5f835a80-8741-11eb-9f9c-9c9498a233ac.png)
**Rotation 1: Obliterate-Obliterate-Blood Strike-Blood Strike**
**Rotation 2: Obliterate-Obliterate-Obliterate**
The 1st rotation is to get my **Death runes**. In this case i converted my **Blood Runes** to **Death Runes** by using **Blood Strike**. **Notice how i wait for approximately 15-20 seconds between my 1st and 2nd rotations- that is to make sure that the runes have enough time to get the maximum grace period**. When i start my **2nd rotation** notice how my **3rd 4th 5th and 6th runes have lower cooldown(7.5 sec)** but **1st and 2nd always start ticking down from 10 sec**. In other words, my rune grace period is working when i am using my **Frost, Unholy and Blood runes** but they are **NOT** working when i am using my **Death Runes.**

### Reproduce Way 2(Reaping):
_video link_ - https://youtu.be/N1h1z-4Jk6k

**Talents used:**
![Screenshot (183)](https://user-images.githubusercontent.com/43633844/111462760-d2420500-8744-11eb-98ac-0ba1485ea0e3.png)
**Rotation 1: Scourge Strike-Scourge Strike-Blood Strike-Blood Strike**
**Rotation 2: Scourge Strike-Scourge Strike-Scourge Strike**
Same as **Reproduce Way 1**. Notice how in the **2nd rotation** when i spend my **Death Runes(1st and 2nd rune)**, they have **higher cooldown(10 sec)** than **3rd 4th 5th and 6th runes(7.5 sec)**

### Reproduce Way 3(Death Rune Mastery):

_video link_ - https://youtu.be/WrR_nHxJXM8

**Talents used:**
![Screenshot (184)](https://user-images.githubusercontent.com/43633844/111463948-2e595900-8746-11eb-88f6-012187effbdb.png)

**Rotation 1: Blood Strike-Blood Strike-Death Strike-Death Strike**
**Rotation 2: Blood Strike-Blood Strike-Death Strike-Death Strike**
The **1st Rotation** is to get my **Death Runes**. In this case i converted my **Frost and Unholy Runes** to **Death Runes** by using **Death Strike**. When i start my **2nd rotation** notice how my **1st and 2nd runes** have **lower cooldown(7.5sec)** but my **3rd 4th 5th and 6th runes** have **10 second cd**.

**Comment by jeyrella** [comment]
Marking this for tester review so we don't push this live, until we are properly sure & agreed on the issue and what is fixed.  

No messing about with skills unless we are **certain** it is not currently correct, and **certain** what it should be.

**Comment by Keader** [comment]
i got a way of reproduce it.
i need janitor help to continue here (server stop build tonight), i will continue tomorrow

**Comment by Keader** [comment]
Test now @hawker357

**Comment by hawker357** [comment]
Just tested on PTR, Death Rune cooldown is now working as expected. 10 second cooldown is only present if an ability is used before the grace period, after the grace period it has a 7.5 second cooldown.
